procedures/unit-testing-utils.ipf

```diff
@@ -893,68 +908,167 @@ static Function NumBytesInUTF8Character(str, byteOffset)
+			return NaN
+		endif
+
+		// no differences in this line
+		start = endmin + strlen(marker1)
+		line += 1
+
+	while(start<strlen(str1) && start<strlen(str2))
+
+	// one string is larger than the other one
+	if(strlen(str1) != strlen(str2))
+		sprintf prefix, "%d:0:%d>", line, start
+		if(strlen(str1) <= start)
+			result.v1 = prefix
+		else
+			end1 = DetectEndOfLine(str1, start, marker1)
+			result.v1 = prefix + EscapeString(str1[start, end1])
+		endif
+		if(strlen(str2) <= start)
+			result.v2 = prefix
+		else
+			end2 = DetectEndOfLine(str2, start, marker2)
+			result.v2 = prefix + EscapeString(str2[start, end2])
+		endif
+		return NaN
+	endif
+
+	Abort("Cannot create diff of equal strings")
+End
+
+static Function/S GetStringWithContext(str, start, diffpos, endpos)
+	variable start, diffpos, endpos
 	string str
 
-	variable bytePos, charNum, numBytes, code
-	variable extPrint, len
-	string format, hexOut
-	string extOut = ""
-
-	Make/FREE/I/U biDiChars = { \
-							char2num("\u200E"), char2num("\u200F"), char2num("\u061C"), char2num("\u202A"), \
-							char2num("\u202D"), char2num("\u202B"), char2num("\u202E"), char2num("\u202C"), \
-							char2num("\u2066"), char2num("\u2067"), char2num("\u2068"), char2num("\u2069")}
-
-	len = strlen(str)
-	if(len)
-		do
-			numBytes = NumBytesInUTF8Character(str, bytePos)
-			code = char2num(str[bytePos, bytePos + numBytes - 1])
-			if(code >= char2num("\u0000") && code <= char2num("\u001F"))
-				// C0 controls
-				extPrint = 1
-			elseif(code >= char2num("\u007F") && code <= char2num("\u009F"))
-				// C1 controls
-				extPrint = 1
+	string strOut
+
+	strOut = EscapeString(str[max(start, diffpos - MAX_STRING_DIFF_CONTEXT), diffpos - 1])
+	strOut += EscapeString(str[diffpos, min(endpos, diffpos + MAX_STRING_DIFF_CONTEXT)])
+	return strOut
+End
+
+static Function GetTextDiffPos(str1, str2)
+	string str1, str2
+
+	variable i
+
+	for(i=0; i < strlen(str1); i += 1)
+		if (!!CmpStr(str1[i], str2[i]))
+			return i
+		endif
+	endfor
+
+	return -1
+End
+
+static Function DetectEndOfLine(str, start, marker)
+	variable start
+	string str, &marker
+
+	variable i
+
+	for(i = start; i < strlen(str); i += 1)
+		if (!CmpStr(str[i], "\r"))
+			if (!CmpStr(str[i + 1], "\n"))
+				// windows line ending
+				marker = "\r\n"
 			else
-				// Bi Directional characters
-				FindValue/V=(code) biDiChars
-				if(V_Value >= 0)
-					extPrint = 1
-				endif
+				// osx line ending
+				marker = "\r"
 			endif
-			format = "%0" + num2str(2 * numBytes) + "x"
-			sprintf hexOut, format, code
-			extOut += hexOut + " "
-			bytePos += numBytes
-		while(bytePos < len)
-	endif
+			return i
+		endif
+		if (!CmpStr(str[i], "\n"))
+			// unix line ending
+			marker = "\n"
+			return i
+		endif
+	endfor
 
-	return SelectString(extPrint, str, ": " + extOut)
+	// no line ending
+	marker = ""
+	return strlen(str)
 End
-#else
-static Function/S AutoExtendedStringOutput(str)
+
+static Function/S EscapeString(str)
```

Comment:
Add a brief description in front of the function that it escapes purely single byte aware.